File: Sep/Day20/implement-router.java
import java.util.*;

class Packet {
    int src, dst, ts;
    Packet(int s, int d, int t) { src = s; dst = d; ts = t; }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Packet)) return false;
        Packet p = (Packet) o;
        return src == p.src && dst == p.dst && ts == p.ts;
    }

    @Override
    public int hashCode() {
        return Objects.hash(src, dst, ts);
    }
}

// ----------- Simple Version -------------
class RouterSimple {
    Deque<Packet> dq;
    Set<Packet> routerHashMap;
    int memoryLimit;

    RouterSimple(int limit) {
        dq = new ArrayDeque<>();
        routerHashMap = new HashSet<>();
        memoryLimit = limit;
    }

    boolean addPacket(int source, int destination, int timestamp) {
        Packet packet = new Packet(source, destination, timestamp);
        if (routerHashMap.contains(packet)) return false;

        if (dq.size() == memoryLimit) forwardPacket();

        dq.addLast(packet);
        routerHashMap.add(packet);
        return true;
    }

    List<Integer> forwardPacket() {
        if (!dq.isEmpty()) {
            Packet packet = dq.pollFirst();
            routerHashMap.remove(packet);
            return Arrays.asList(packet.src, packet.dst, packet.ts);
        }
        return new ArrayList<>();
    }

    int getCount(int destination, int startTime, int endTime) {
        int count = 0;
        for (Packet p : dq) {
            if (p.dst == destination && startTime <= p.ts && p.ts <= endTime) count++;
        }
        return count;
    }
}

// ----------- Optimized Version -------------
class RouterOptimized {
    Deque<Packet> dq;
    Set<Packet> routerHashMap;
    Map<Integer, TreeMap<Integer, Integer>> destinationMap; // multiset simulation
    int memoryLimit;

    RouterOptimized(int limit) {
        dq = new ArrayDeque<>();
        routerHashMap = new HashSet<>();
        destinationMap = new HashMap<>();
        memoryLimit = limit;
    }

    boolean addPacket(int source, int destination, int timestamp) {
        Packet packet = new Packet(source, destination, timestamp);
        if (routerHashMap.contains(packet)) return false;

        if (dq.size() == memoryLimit) forwardPacket();

        dq.addLast(packet);
        routerHashMap.add(packet);

        destinationMap.putIfAbsent(destination, new TreeMap<>());
        TreeMap<Integer, Integer> map = destinationMap.get(destination);
        map.put(timestamp, map.getOrDefault(timestamp, 0) + 1);

        return true;
    }

    List<Integer> forwardPacket() {
        if (!dq.isEmpty()) {
            Packet packet = dq.pollFirst();
            routerHashMap.remove(packet);

            TreeMap<Integer, Integer> map = destinationMap.get(packet.dst);
            map.put(packet.ts, map.get(packet.ts) - 1);
            if (map.get(packet.ts) == 0) map.remove(packet.ts);
            if (map.isEmpty()) destinationMap.remove(packet.dst);

            return Arrays.asList(packet.src, packet.dst, packet.ts);
        }
        return new ArrayList<>();
    }

    int getCount(int destination, int startTime, int endTime) {
        if (!destinationMap.containsKey(destination)) return 0;
        TreeMap<Integer, Integer> map = destinationMap.get(destination);
        int count = 0;
        for (var entry : map.subMap(startTime, true, endTime, true).values()) {
            count += entry;
        }
        return count;
    }
}

File: Sep/Day20/implement-router.py
from collections import deque
from typing import List

class Router:

    def __init__(self, memoryLimit: int):
        self.dq = deque()               # queue of packets
        self.routerHashMap = set()      # fast lookup to avoid duplicates
        self.memoryLimit = memoryLimit

    def addPacket(self, source: int, destination: int, timestamp: int) -> bool:#O(1)
        packet = (source, destination, timestamp)  # use tuple
        if packet in self.routerHashMap:
            return False

        if len(self.dq) == self.memoryLimit:
            self.forwardPacket()

        self.dq.append(packet)
        self.routerHashMap.add(packet)
        return True

    def forwardPacket(self) -> List[int]:#O(1)
        if self.dq:
            packet = self.dq.popleft()
            self.routerHashMap.remove(packet)
            return list(packet)
        return []

    def getCount(self, destination: int, startTime: int, endTime: int) -> int: #O(n)
        count = 0
        for src, dst, ts in self.dq:
            if dst == destination and startTime <= ts <= endTime:
                count += 1
        return count


class Router:

    def __init__(self, memoryLimit: int):
        self.dq = deque()                     # queue of packets
        self.routerHashMap = set()            # fast duplicate check
        self.memoryLimit = memoryLimit
        self.destinationMap = {}              # dest -> SortedList of timestamps

    def addPacket(self, source: int, destination: int, timestamp: int) -> bool:#O(1)
        packet = (source, destination, timestamp)
        if packet in self.routerHashMap:
            return False

        if len(self.dq) == self.memoryLimit:
            self.forwardPacket()

        self.dq.append(packet)
        self.routerHashMap.add(packet)

        if destination not in self.destinationMap:
            self.destinationMap[destination] = SortedList()
        self.destinationMap[destination].add(timestamp)   # O(log n)

        return True

    def forwardPacket(self) -> List[int]:#O(logN)
        if self.dq:
            packet = self.dq.popleft()
            self.routerHashMap.remove(packet)

            src, dst, ts = packet
            self.destinationMap[dst].remove(ts)   # O(log n)
            if not self.destinationMap[dst]:
                del self.destinationMap[dst]

            return list(packet)
        return []

    def getCount(self, destination: int, startTime: int, endTime: int) -> int:#O(logN)
        if destination not in self.destinationMap:
            return 0

        timestamps = self.destinationMap[destination]
        left = timestamps.bisect_left(startTime)
        right = timestamps.bisect_right(endTime)
        return right - left